graph/ForceDirectedGraph/ForceDirectedGraph.tsx

```diff
@@ -0,0 +1,264 @@
+import React, { useEffect, useState, useRef } from "react";
+import { View, Dimensions, Text, StyleSheet } from "react-native";
+import Svg, { Circle, Line, G } from "react-native-svg";
+import Graph, { Link, Node } from "../Graph";
+import { fruchtermanReingold } from "../graph-algorithms/fruchtermanReingold";
+import {
+  PanGestureHandler,
+  PinchGestureHandler,
+  State,
+  GestureHandlerRootView,
+} from "react-native-gesture-handler";
+
+const SHORT_PRESS_DURATION = 100;
+
+const DEFAULT_NODE_SIZE = 10;
+const DEFAULT_NODE_SIZE_INCREMENT = 2;
+
+const DEFAULT_NODE_COLOR = "blue";
+const CONSTRAINED_NODE_COLOR = "red";
+
+const DEFAULT_LINK_COLOR = "black";
+
+const DEFAULT_CLICKED_NODE_ID = "";
+
+const MAX_ITERATIONS = 1000;
+
+const ForceDirectedGraph: React.FC<{
+  graph: Graph;
+  constrainedNodeId: string;
+}> = ({ graph, constrainedNodeId }) => {
+  const [nodes, setNodes] = useState<Node[]>([]);
+  const [links, setLinks] = useState<Link[]>([]);
+  const [sizes, setSizes] = useState<Map<string, number>>(new Map());
+  const [load, setLoad] = useState<boolean>(false);
+  const [dragEnabled, setDragEnabled] = useState(true); // Toggle drag
+  const [clickedNodeID, setClickedNodeID] = useState<string>(
+    DEFAULT_CLICKED_NODE_ID
+  ); // Node ID of clicked node
+  const [totalOffset, setTotalOffset] = useState({ x: 0, y: 0 });
+  const [scale, setScale] = useState(1);
+  const [lastScale, setLastScale] = useState(1); // Add state to keep track of last scale
+
+  const pressStartRef = useRef(0);
+
+  const screenWidth = Dimensions.get("window").width;
+  const screenHeight = Dimensions.get("window").height;
+
+  const panRef = useRef(null);
+  const pinchRef = useRef(null);
+
+  useEffect(() => {
+    const initialLinks = graph.getLinks();
+    const initialNodes = graph.getNodes();
+    const initialSizes = nodesSizesInit([...initialLinks]);
+    setLinks([...initialLinks]);
+
+    setNodes(
+      fruchtermanReingold(
+        [...initialNodes],
+        initialLinks,
+        constrainedNodeId,
+        screenWidth,
+        screenHeight,
+        MAX_ITERATIONS
+      )
+    );
+    setSizes(initialSizes);
+    setLoad(true);
+    return () => {
+      setLinks([]);
+      setNodes([]);
+      setSizes(new Map());
+      setLoad(false);
+    };
+  }, [graph, constrainedNodeId, screenWidth, screenHeight]);
```

Comment:
screenWidth and screenHeight normally should not change. I think you can safely remove them from the sensitivity list